ksql-metastore/src/test/java/io/confluent/ksql/metastore/model/KeyFieldTest.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.metastore.model;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.testing.EqualsTester;
+import io.confluent.ksql.util.KsqlConfig;
+import java.util.Optional;
+import org.apache.kafka.connect.data.Field;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public class KeyFieldTest {
+
+  private static final Schema SCHEMA = SchemaBuilder.struct()
+      .field("field0", Schema.OPTIONAL_STRING_SCHEMA)
+      .field("field1", Schema.OPTIONAL_BYTES_SCHEMA)
+      .build();
+
+  private static final KsqlConfig LEGACY_CONFIG = new KsqlConfig(
+      ImmutableMap.of(KsqlConfig.KSQL_USE_LEGACY_KEY_FIELD, true)
+  );
+
+  private static final KsqlConfig LATEST_CONFIG = new KsqlConfig(
+      ImmutableMap.of()
+  );
+
+  private static final Field RANDOM_FIELD =
+      new Field("won't find me anywhere", 0, Schema.OPTIONAL_STRING_SCHEMA);
+
+  private static final Field SCHEMA_FIELD = SCHEMA.fields().get(0);
+
+  @Rule
+  public final ExpectedException expectedException = ExpectedException.none();
+
+  @Test
+  public void shouldImplementHashCodeAndEqualsProperly() {
+    final Optional<String> keyField = Optional.of("key");
+    final Optional<Field> legacy = Optional.of(SCHEMA.fields().get(0));
+
+    new EqualsTester()
+        .addEqualityGroup(KeyField.of(keyField, legacy), KeyField.of(keyField, legacy))
+        .addEqualityGroup(KeyField.of(Optional.empty(), legacy))
+        .addEqualityGroup(KeyField.of(keyField, Optional.empty()))
```

Comment:
nit: missing double empty

Reply:
I'll take a look.

Reply:
I don't think it's necessary, but I've added it anyway.